src/emsarray/operations/cache.py

```diff
@@ -0,0 +1,108 @@
+    hash : hashlib-style hash instance
+        The hash instance to update with the given attribute dict.
+        This must follow the interface defined in :mod:`hashlib`.
+    attributes: dict
+        Expects a marshal compatible dictionary
+    """
+    # Prepend the marshal encoding version
+    marshal_version = 4
+    hash_int(hash, marshal_version)
+    # Specify marshal encoding version when serialising
+    attribute_dict_marshal_bytes = marshal.dumps(attributes, marshal_version)
+    # Prepend the number of attributes
+    hash_int(hash, len(attributes))
+    # Prepend the size of the pickled attributes
+    hash_int(hash, len(attribute_dict_marshal_bytes))
+    hash.update(attribute_dict_marshal_bytes)
+
+
+def hash_string(hash: "hashlib._Hash", value: str) -> None:
+    """
+    Updates the provided hash with with a utf-8 encoded byte representation of the provided string.
+
+    Parameters
+    ----------
+    hash : hashlib-style hash instance
+        The hash instance to update with the given attribute dict.
+        This must follow the interface defined in :mod:`hashlib`.
+    attributes: str
+        Expects a string that can be encoded in utf-8
+    """
+    # Prepend the str length
+    hash_int(hash, len(value))
+    hash.update(value.encode('utf-8'))
+
+
+def hash_int(hash: "hashlib._Hash", value: int) -> None:
+    """
+    Updates the provided hash with an encoded byte representation of the provided int.
+
+    Parameters
+    ----------
+    hash : hashlib-style hash instance
+        The hash instance to update with the given attribute dict.
+        This must follow the interface defined in :mod:`hashlib`.
+    attributes: int
+        Expects an int that can be represented in a numpy int32
+    """
+    # Prepend the int bit length
+    hash.update(numpy.int32(value.bit_length()).tobytes())
+    hash.update(numpy.int32(value).tobytes())
+
+
+def make_cache_key(dataset: xarray.Dataset, hash: "hashlib._Hash | None" = None) -> str:
+    """
+    Generate a key suitable for caching data derived from the geometry of a dataset.
+
+    Parameters
+    ----------
+    dataset : xarray.Dataset
+        The dataset to generate a cache key from
+    hash : hash instance
+        An instance of a hashlib hash class
+        Defaults to `hashlib.blake2b`, which is secure enough and fast enough for most purposes.
+        The hash algorithm does not need to be cryptographically secure,
+        so faster algorithms such as `xxhash` can be swapped in if desired.
+
+    Returns
+    -------
+    cache_key : str
+        A string suitable for use as a cache key.
+        The string will be safe for use as part filename if data are to be cached to disk.
+
+    Notes
+    -----
+    The cache key will depend on the Convention class,
+    the emsarray version, and a hash of the geometry of the dataset.
+    The specific structure of the cache key may change between emsarray and python versions,
+    and should not be relied upon.
```

Comment:
The documentation is automatically built as part of the pull request CI process, and linked from the build checks section. You can find the built documentation for this PR over at https://emsarray--158.org.readthedocs.build/en/158/. You can also compile the documentation locally with:

```shell
$ cd emsarray
$ pip install -e .[docs]
$ make -C docs live
```

The documentation will be available locally at http://127.0.0.1:5500/

This cache module is not included in the documentation. Using the other operations modules as a guide, can you please add documentation for this module? You'll need to create a `docs/api/operations/cache.rst` file and then add a reference to that to `docs/api/operations/index.rst`. I think we should only expose the `make_cache_key()` function in this module, while the `hash_attributes()`, `hash_string()`, and `hash_int()` helper methods can be kept private and not included in the generated documentation.

Once this modules documentation is being built, please review the generated documentation to ensure it reads well. I noticed that the docstring I wrote in my original feature request is missing a trailing "." for the first sentence of the hash parameter, for example.